### PR TITLE
Add key validation in dh generate key

### LINKED ISSUE #21781: Weak key in dh generate key
Version: OpenSSL3.x
There is no public key validation  in DH_generate_key, and weak public key may generated ,this will result weak DH key.

### PR DESCRIPTION
**Pull request by KanPlus**
Fixes #21781

### CODE REVIEW

**Comment by kroeckx** [comment]
Can you clarify why you think this check is needed?

The public key is generated from the private key using `g^priv_key mod p`.
The check that you add now checks that pubkey is in the range `[2, p - 1]`, and if we have q, that `pub_key^q == 1 mod p`.

Are you worried that the public key might become 1? At least in the cases where we have q, ossl_ffc_generate_private_key() is used, and it will always be in the `[2, p - 1]` range because it's already been tested. You could argue that in the other case, there is a very small chance that 1 is generated.

**Comment by KanPlus** [comment]
> 

The private key could not be valided in case of no dh->params.q , and the public key could be valided.

**Comment by KanPlus** [comment]
> The private key could not be valided in case of no dh->params.q , and the public key could be valided.

The private key could not be valided in case of no dh->params.q , and the public key could be valided.

**Comment by slontis** [comment]
I don't think this PR is needed., If you need to do key checks after generation then something is very wrong..
SP800-56A and 56B specify that key checks are not required if the key generation uses a compliant algorithm

**Comment by KanPlus** [comment]
> SP800-56A

It is better to regenerate the key in case of invalid key.
"5.6.2.3 Public Key Validation Routines" define the check logic in "SP800-56A".

**Comment by slontis** [comment]
It depends on how you interpret this:

For either a static key pair or an ephemeral key pair: The owner generates the key
pair as specified in Section 5.6.1 and performs a full public-key validation or an
equivalent procedure as part of its generation process (see Section 5.6.2.3.1 for FFC,
and Section 5.6.2.3.3 for ECC);

i.e. If the process of keygen already satisfies these constraints then the check is pointless.

**Comment by KanPlus** [comment]
> keygen

Yes, this PR is aim to add the missed validation in case of  no params q when generate dh key pair.

**Comment by kroeckx** [comment]
When we don't have q, we don't use SP800-56A. Assuming a 2048 bit p, I think that gets you a 1 in 2^2047 chance of generating a public key of 1, which most people will ignore. As far as I know, SP800-56A has a loop for the case that it does generate a public key of 1.

File: crypto/dh/dh_key.c [hunk 1]
```diff
@@ -342,22 +351,21 @@ static int generate_key(DH *dh)
...
-                 */
+                * For FFC FIPS 186-4 keygen
+                * security strength s = 112,
+                * Max Private key size N = len(q)
+                */
                 if (!ossl_ffc_generate_private_key(ctx, &dh->params,
-                                                   BN_num_bits(dh->params.q),
-                                                   MIN_STRENGTH,
-                                                   priv_key))
+                                                BN_num_bits(dh->params.q),
```
Comment (slontis) [inline crypto/dh/dh_key.c:359]: indenting is not right here..

File: crypto/dh/dh_key.c [hunk 2]
```diff
@@ -310,30 +310,39 @@ static int generate_key(DH *dh)
...
+                        /* clear bit 0, since it won't be a secret anyway */
+                        if (!BN_clear_bit(priv_key, 0))
+                            goto err;
+                    }
+                    if (!ossl_dh_generate_public_key(ctx, dh, priv_key, pub_key))
+                        goto err;
+
+                    if (DH_check_pub_key_ex(dh, pub_key))
+                        break;
+                } while (1);
```
Comment (slontis) [inline crypto/dh/dh_key.c:345]: Is there any possible way this could now be an infinite loop?

Comment (KanPlus) [reply crypto/dh/dh_key.c:358]: Fixed the possible infinite loop.

File: crypto/dh/dh_key.c (hunk 1 above)
Comment (KanPlus) [reply crypto/dh/dh_key.c]: Fixed.

**Comment by t8m** [comment]
IMO we do not need this.

> When we don't have q, we don't use SP800-56A. Assuming a 2048 bit p, I think that gets you a 1 in 2^2047 chance of generating a public key of 1, which most people will ignore. As far as I know, SP800-56A has a loop for the case that it does generate a public key of 1.

I agree with Kurt. IMO we do not need and want this. At most we should use ossl_ffc_validate_public_key_partial() instead. Otherwise we will be doing another modular exponentiation making the key generation slower.

**Comment by KanPlus** [comment]
> 

Yes, the upper of private key could not be limited  with "q-1", and this will  increase the chance. this will help to ingnore the weak key.

**Comment by KanPlus** [comment]
> 

This 

> IMO we do not need this.
> 
> > When we don't have q, we don't use SP800-56A. Assuming a 2048 bit p, I think that gets you a 1 in 2^2047 chance of generating a public key of 1, which most people will ignore. As far as I know, SP800-56A has a loop for the case that it does generate a public key of 1.
> 
> I agree with Kurt. IMO we do not need and want this. At most we should use ossl_ffc_validate_public_key_partial() instead. Otherwise we will be doing another modular exponentiation making the key generation slower.

The chance maybe increase as the  upper of private key could not be limited with "q-1", and the logic is same with  ossl_ffc_validate_public_key_partial as the another modular exponentiation as q not present.

**Comment by KanPlus** [comment]
> When we don't have q, we don't use SP800-56A. Assuming a 2048 bit p, I think that gets you a 1 in 2^2047 chance of generating a public key of 1, which most people will ignore. As far as I know, SP800-56A has a loop for the case that it does generate a public key of 1.

Yes, the upper of private key could not be limited with "q-1", and this will increase the chance. this will help to ingnore the weak key.

**Comment by KanPlus** [comment]
> It depends on how you interpret this:
> 
> For either a static key pair or an ephemeral key pair: The owner generates the key pair as specified in Section 5.6.1 and performs a full public-key validation or an equivalent procedure as part of its generation process (see Section 5.6.2.3.1 for FFC, and Section 5.6.2.3.3 for ECC);
> 
> i.e. If the process of keygen already satisfies these constraints then the check is pointless.

Yes, this PR is aim to add the missed validation in case of no params q when generate dh key pair.

**Comment by bernd-edlinger** [comment]
But you could also have other weak keys e.g. simple powers of 2, which might happen
if you accidentally choose the private key `(p-1)/2 + n` will get public key `g^n` or `p-g^n`
and those are slightly more likely than `1` or `p-1` alone.
And those will not make the proposed check fail?

**Comment by bernd-edlinger** [comment]
So how about that instead?
```
diff --git a/crypto/dh/dh_key.c b/crypto/dh/dh_key.c
index 328780b9a9..a352ec355a 100644
--- a/crypto/dh/dh_key.c
+++ b/crypto/dh/dh_key.c
@@ -320,7 +320,7 @@ static int generate_key(DH *dh)
                 if (dh->length != 0
                     && dh->length >= BN_num_bits(dh->params.p))
                     goto err;
-                l = dh->length ? dh->length : BN_num_bits(dh->params.p) - 1;
+                l = dh->length ? dh->length : BN_num_bits(dh->params.p) - 2;
                 if (!BN_priv_rand_ex(priv_key, l, BN_RAND_TOP_ONE,
                                      BN_RAND_BOTTOM_ANY, 0, ctx))
                     goto err;
```
would make the private key not in the range `(p-1)/2` ... `(p-1)/2+n`
and there is no chance that a public key can be 1, 2, 4, 8 etc. ?

**Comment by KanPlus** [comment]
> So how about that instead?
> 
> ```
> diff --git a/crypto/dh/dh_key.c b/crypto/dh/dh_key.c
> index 328780b9a9..a352ec355a 100644
> --- a/crypto/dh/dh_key.c
> +++ b/crypto/dh/dh_key.c
> @@ -320,7 +320,7 @@ static int generate_key(DH *dh)
>                  if (dh->length != 0
>                      && dh->length >= BN_num_bits(dh->params.p))
>                      goto err;
> -                l = dh->length ? dh->length : BN_num_bits(dh->params.p) - 1;
> +                l = dh->length ? dh->length : BN_num_bits(dh->params.p) - 2;
>                  if (!BN_priv_rand_ex(priv_key, l, BN_RAND_TOP_ONE,
>                                       BN_RAND_BOTTOM_ANY, 0, ctx))
>                      goto err;
> ```
> 
> would make the private key not in the range `(p-1)/2` ... `(p-1)/2+n` and there is no chance that a public key can be 1, 2, 4, 8 etc. ?

The p was not present, so the verification that depends on p could not be performed. The verification here complies with the best-effort principle.

**Comment by bernd-edlinger** [comment]
But since p is always a strong prime number we know by Fermat's little theorem
that `g^(p-1) == 1 (mod p)`  and the only possible solution for `g^x == 1 mod p`
is `x == p-1` and `x == (p-1)/2`
Therefore the check is superfluous if we simply use a private key less than `(p-1)/2' as suggested.

**Comment by KanPlus** [comment]
> But since p is always a strong prime number we know by Fermat's little theorem that `g^(p-1) == 1 (mod p)` and the only possible solution for `g^x == 1 (mod p)` is `x == p-1` and `x == (p-1)/2` Therefore the check is superfluous if we simply use a private key less than `(p-1)/2` as suggested.

As the p not present, so the private key could not be limited with  (p-1)/2; so this PR add the  public key validation with the present q.

**Comment by t8m** [comment]
> As the p not present, so the private key could not be limited with (p-1)/2; so this PR add the public key validation with the present q.

I am not sure what you're talking here about - p not present is impossible - this would not be a key with valid parameters.

**Comment by kroeckx** [comment]
It's my understanding that DH groups do not need to be a safe prime, and there are various examples of that in the RFCs. OpenSSL even has support for creating DSA-style DH parameters, where p-1 is a multiple of q. So even for the cases we have a q, we don't know if it's a safe prime or not. It's my understanding that only for the case of  built-in parameters like the ffdhe ones we can be sure it's a safe prime.

**Comment by KanPlus** [comment]
> > As the p not present, so the private key could not be limited with (p-1)/2; so this PR add the public key validation with the present q.
> 
> I am not sure what you're talking here about - p not present is impossible - this would not be a key with valid parameters.

Sorry, it is q that not present here.

**Review by jamiegreen7 — CHANGES_REQUESTED** [review]
Change private key

**Pull request closed without merging**